src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -597,21 +454,48 @@ private ValueTask WriteSingleChunk<TIOAdapter>(TIOAdapter writeAdapter, ReadOnly
             if (t.IsCompletedSuccessfully)
             {
                 ArrayPool<byte>.Shared.Return(rentedBuffer);
-                FinishWrite();
                 return t;
             }
             else
             {
-                return CompleteAsync(t, rentedBuffer);
+                return CompleteWriteAsync(t, rentedBuffer);
             }
 
-            async ValueTask WaitForWriteIOSlot(TIOAdapter wAdapter, Task lockTask, ReadOnlyMemory<byte> buff)
+            async ValueTask WaitAndWriteAsync(Task t, byte[] bufferToReturn)
             {
-                await lockTask.ConfigureAwait(false);
-                await WriteSingleChunk(wAdapter, buff).ConfigureAwait(false);
+                try
+                {
+                    await t.ConfigureAwait(false);
```

Comment:
A lot of this logic appears to be duplicated with the caller. If that's true, since we expect this to be rare, could we just await a recursive call back to the caller? If there's any concern about stack diving from the recursion, we could queue the call. It'll end up being more expensive when it occurs, but it would seem to simplify things (feel free to tell me I'm wrong :) and this is supposed to be very rare.

Reply:
I updated the logic as you suggested @stephentoub. While working on this I realized we don't have a good way how to test this.  Getting one renegotiation is easy (either with TLS1.3 or with external server) but to do that again is not easy until we have some way how to trigger it in loop with concurrent writes,

Reply:
I wish we had an API in SslStream (server-side) to trigger a renegotiation. :)  Perhaps it is time to add one.

And while we wait for an API approval, we could perhaps make this API 'internal' and use reflection in our tests to invoke it.

Reply:
> I wish we had an API in SslStream (server-side) to trigger a renegotiation. :) Perhaps it is time to add one. And while we wait for an API approval, we could perhaps make this API 'internal' and use reflection in our tests to invoke it.

I agree.  However, it seems likely introducing that is going to require some re-work in the locking scheme, since the current scheme in this PR expects that renegotation can only be triggered by reads and relies on that for correctness.  I think we should get this change in, in order to fix the known existing problems and unblock other work, and then introducing a new method (even if internal at first to help with testing) sounds like a good idea.

Reply:
It is on my TODO list. I think I know how to do that at least for schannel and OpenSSL. Also note, that for TLS13 we go through this path even if renegotiation does not exist in that version. When I talked to experts, it seems like that particular status code is used anytime when something needs to be sent to `lsass` for further processing.